content/post/2020-01-02-cran-error/index.Rmd

```diff
@@ -0,0 +1,106 @@
+
+:sparkles: Tip \#2 :sparkles: *search the net!*
+
+## R and docker
+
+<!-- r-hub and rocker - which is good for what -->
+There are two services that are worth mentioning:
+
+- [rhub](https://r-hub.github.io/rhub/), which is an R package for using the [R-hub package builder](https://docs.r-hub.io/#package-builder),
+- and the [Rocker project](https://www.rocker-project.org/), which distributes various Docker images useful for R and RStudio testing
+
+If you are a maintainer of an R package and don't want to install Docker locally, you will find the `rhub` package very useful, as you can push your package to the server for testing with one command! Check out e.g., [the package docs](https://r-hub.github.io/rhub/)
+
+If you want to test more locally, and importantly, if you want to test visualizations, GUI, or specific behavior of the code in RStudio, you should take a look on the `Rocker project`. They provide easy-to-run Docker images with pre-installed R and RStudio, which you can play with through a browser window.
+
+## R-hub locally
+
+In my case, I am not registered as the main maintainer of Haplin, and I wanted to test locally, so I decided to use one of the `rhub` functions, [`local_check_linux()`](https://r-hub.github.io/rhub/reference/local_check_linux.html). This seemed like a very easy way to run locally tests on a platform that is exactly the same as CRAN uses! And it is... only not in my case ;)
+
+After getting the same error for the nth time, I decided to get help [at the source](https://github.com/r-hub/rhub/issues/322). I tried several things, including digging into the `rhub` code locally, but nothing worked. Long story short - my Docker installation by default gave no access to internet to the containers it launched.
+
+:sparkles: Tip \#3 :sparkles: *ask for help other people!* (they don't bite, especially not via e-mails)
+
+## Script for semi-manual checks
+
+Another couple of days and I talked with my husband, who showed me a script he used in one of his projects. [The script](https://bitbucket.org/Grantlab/bio3d/src/master/ver_devel/util/run_dockercheck.sh) uses a local Docker installation, fetches one of [the images hosted by R-hub](https://github.com/r-hub/rhub-linux-builders#rhub-linux-builders), and runs tests locally. *Bingo!* As this was the second time I tried using Docker, I needed some attempts to adjust the script to my purposes, but it finally worked!
+
+
+## Conclusion
+
+Since I could now reproduce the exact error and I knew roughly which parts of the code caused it, I fixed the problems relatively fast. The main issue, as mentioned above, was with the identification of a class of the objects, which I've re-written now to ``is( obj, "myClass")`` instead, which is also a recommended way of checking class memebership.
```

Comment:
```suggestion
Since I could now reproduce the exact error and I knew roughly which parts of the code caused it, I fixed the problems relatively fast. The main issue, as mentioned above, was with the identification of a class of the objects, which I've re-written now to ``is( obj, "myClass")`` instead, which is also a recommended way of checking class membership.
```